### PR TITLE
Generic cleanup Metrics and Losses

### PR DESCRIPTION
**Pull request by JimClarke5**
This PR Is a reissue of the Generic Cleanup PR #203, removing commits in the branch that are already in `master`.

This PR effects both `losses` and `metrics`.

This PR cleans up generic parameters to cut down on the number of unique parameters to at most 1 where possible.
For example:
`public abstract class Metric<U extends TNumber, T extends TNumber> {`
becomes
`public abstract class Metric<T extends TNumber> {`

and

```
@Override
  public <V extends TNumber> List<Op> updateStateList(Operand<U> values, Operand<V> sampleWeights) {
```
Changes to:
```
 public List<Op> updateStateList(
      Operand<? extends TNumber> values, Operand<? extends TNumber> sampleWeights)

### CODE REVIEW

**Comment by karllessard** [comment]
Hi @JimClarke5 , there are some javadoc issues in this PR, please take a look: https://github.com/tensorflow/java/pull/204/checks?check_run_id=1825661890

**Comment by karllessard** [comment]
BTW, do you have any clue why the javadoc has been reformatted? It looks like it was already conform before, I don't know why your IDE plugin would reformat these, can you please just double-check that you use your original settings?

**Comment by JimClarke5** [comment]
@karllessard  I fixed all the JavaDoc issues under `framework` except for the `data` package. 
I have no clue why the format was messed up, but I applied IntelliJ Google format on the `optimizers`, `losses`, `metrics` and `utils` directories.

File: tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/SparseCategoricalCrossentropy.java [hunk 1]
```diff
@@ -21,15 +21,16 @@
 import org.tensorflow.op.Ops;
 import org.tensorflow.types.family.TNumber;
 
+import static org.tensorflow.framework.utils.CastHelper.cast;
+
 /**
  * A metric that computes the sparse categorical cross-entropy loss between true labels and
- * predicted labels.
+ * predicted labels. \
```
Comment (karllessard) [inline tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/SparseCategoricalCrossentropy.java:28]: Is this backslash really there for a reason?

**Review by karllessard — APPROVED** [review]
Great work @JimClarke5 , that simplifies really a lot some of the APIs. I just left a really-unimportant comment about something that looks like a trailing character, please confirm and/or fix and we are good to merge

**Comment by karllessard** [comment]
Ok nevermind, I've remove the trailing character directly from your branch, this is now ready to merge, thanks!

**Review by karllessard — APPROVED** [review]
